time.go

```diff
@@ -0,0 +1,96 @@
+package null
+
+import (
+	"database/sql/driver"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"time"
+)
+
+// Time is a nullable time.Time. It supports SQL and JSON serialization.
+// It will marshal to null if null.
+type Time struct {
+	Time  time.Time
+	Valid bool
+}
+
+// Scan implements the Scanner interface.
+func (t *Time) Scan(value interface{}) error {
+	t.Time, t.Valid = value.(time.Time)
+	return nil
+}
+
+// Value implements the driver Valuer interface.
+func (t Time) Value() (driver.Value, error) {
+	if !t.Valid {
+		return nil, nil
+	}
+	return t.Time, nil
+}
+
+// NewTime creates a new Time.
+func NewTime(t time.Time, valid bool) Time {
+	return Time{
+		Time:  t,
+		Valid: valid,
+	}
+}
+
+// TimeFrom creates a new Time that will always be valid.
+func TimeFrom(t time.Time) Time {
+	return NewTime(t, true)
+}
+
+// TimeFromPtr creates a new Time that will be null if t is nil.
+func TimeFromPtr(t *time.Time) Time {
+	if t == nil {
+		var ti time.Time
+		return NewTime(ti, false)
+	}
+	return NewTime(*t, true)
+}
+
+// MarshalJSON implements json.Marshaler.
+// It will encode null if this time is null.
+func (t Time) MarshalJSON() ([]byte, error) {
+	if !t.Valid {
+		return []byte("null"), nil
+	}
+	return json.Marshal(t.Time)
+}
+
+// UnmarshalJSON implements json.Unmarshaler.
+// It supports time, string and null input.
+func (t *Time) UnmarshalJSON(data []byte) error {
+	var err error
+	var v interface{}
+	json.Unmarshal(data, &v)
+	switch x := v.(type) {
+	case time.Time:
+		t.Time = x
+	case string:
+		t.Time, err = time.Parse(time.RFC3339, x)
+	case nil:
+		t.Valid = false
+		return nil
+	default:
```

Comment:
A few things:
* do we want to use `err = t.Time.UnmarshalJSON([]byte(x))` or `err = json.Umarshal([]byte(x), &t.Time)`? I guess the more general question is what's the difference between `anything.UnmarshalJSON(data)` vs `json.Unmarshal(data, &anything)`?
* In the case of `map[string]interface` I think panicking without being able to return the error is definitely less than ideal. I think it would be better to check types (rather than asserting) so we can gracefully return the error.
* Good catch with unmarshalling into `time.Time`.
* I totally agree that we should the JSON error message.